### ISSUE TITLE
Issue to get data using php8 client which stored from php7 client

### ISSUE THREAD

**Issue by piyush-bansal1989** [aerospike/php-client#48]
@vmsachin / @khaf  : I'm encountering an issue when retrieving data stored by a PHP 7 client using a PHP 8 client. The vice versa operation (retrieving data stored by PHP 8 using PHP 7) works correctly. The data format for storage is consistent.

Could you please assist me in resolving this issue? Please find the sample data attached.

**PHP7 client store data which not able retrieving from php8 client**

`aql> select * from ns_amcat_datastore.amcat_auth_tokens where PK='1d6eed27fdc8b35d46ed7d48';
+----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+-------+
| PK                         | content                                                                                   | {ttl} | {gen} |
+----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+-------+
| "1d6eed27fdc8b35d46ed7d48" | MAP('{"authDetails":{"apiSecret":"**********", "apiToken":"1d6eed27fdc8b35d46ed7d48", "createdOn":"2024-12-20 10:34:38", "expiresOn":"2024-12-21 10:34:38", "isActive":1, "userName":"******", "userType":"******"}, "configEntities":{"A | 85586 | 5     |
+----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+-------+
1 row in set (0.000 secs)`

**Below data stored using php8 client which able to retrieving in php7 client**

`aql> select * from ns_amcat_datastore.amcat_auth_tokens where PK='f8aa10dac9111e70b4b09de2';
+----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+-------+
| PK                         | content                                                                                   | {ttl} | {gen} |
+----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+-------+
| "f8aa10dac9111e70b4b09de2" | MAP('{"authDetails":{"apiSecret":"*********", "apiToken":"f8aa10dac9111e70b4b09de2", "createdOn":"2024-12-19 18:24:32", "expiresOn":"2024-12-20 18:24:32", "isActive":1, "userName":"********", "userType":"********"}, "configEntities":{"A | 27454 | 4     |
+----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+-------+`

We require fix for this issue as we move traffic from old to new one distributed way.

Waiting for your revert thanks

**Comment by khaf** [comment]
Can you provide a PHP 7 snippet that reproduces the issue? The information you have provided is not enough for me to reproduce the issue unfortunately.

What is the error you get from the PHP8 client?

**Comment by piyush-bansal1989** [comment]
@khaf / @vmsachin: PFA the sample data, code snippet for both php7 and php8 clients.
[code-snippet-files.zip](https://github.com/user-attachments/files/18218072/code-snippet-files.zip)

First run the old-aerospike -client which put and get data from aerospike. 
Run new-aerospike-client which only get data from aerospike from same namspace, set and key

I didn't get any error from php8 client recieved the following kind of response.
Please also see the ttl always recieved 1 in response but the actual the ttl is showing in the aerospike server.
`Aerospike\Record Object
(
    [bins] => Array
        (
            [content] => Array
                (
                )

        )

    [ttl] => 1
    [key] => Aerospike\Key Object
        (
            [digest] => f86b5c15e94856ec5a7144ec2800cf976c85a992
            [setname] => test
            [value] => testing
            [namespace] => ns_amcat_datastore
            [partition_id] => 3064
        )

    [expiration] => Aerospike\Expiration Object
        (
        )

    [generation] => 28
)`

**Comment by piyush-bansal1989** [comment]
@khaf / @vmsachin : I've found the issue, The PHP 8 Aerospike client cannot reliably retrieve data written by the PHP 7 Aerospike client when that data is stored in array/list format and contains true/false/null values.

- The PHP 8 aerospike client handles true, false, and null values differently, converting them to hexadecimal representations (e.g., true: 62 3A 31 3B, false: 62 3A 30 3B, null: 4E 3B) when storing data in array/list format.
- Conversely, when the PHP 7 aerospike client retrieves data stored by the PHP 8 aerospike client, it incorrectly modifies the values: true becomes 1, false becomes 0, and null value keys are removed.

Could you please fix this issue? So suggest some other solution.

**Comment by piyush-bansal1989** [comment]
Happy New Year, @khaf and @vmsachin!
I hope you have had a restful holiday season. Could you please review this issue at your earliest convenience?

**Comment by khaf** [comment]
Hi @piyush-bansal1989 , happy new year! Sorry I have been looking into this issue for the last few days but I have not updated this ticket. I need until early next week to give a proper update. 

ps. @vmsachin is not working on this project anymore, you can ping me instead.

**Comment by piyush-bansal1989** [comment]
@khaf: Do you have any update on this issue.

**Comment by piyush-bansal1989** [comment]
@khaf : I would request you to share some update on this issue.

**Comment by piyush-bansal1989** [comment]
@khaf : I hope you doing well! , will you please share the update on this issue?

**Comment by khaf** [comment]
@piyush-bansal1989 Hi Piyush, sorry for my late reply. I am working on this issue still, it seems like a serialization compatibility issue between our old C client and the newer clients. I'll keep you updated on what exactly is the issue and how we will deal with it in a couple of days.

**Comment by piyush-bansal1989** [comment]
Hi @khaf  Is there any update on it, our product upgrade peding on it.

**Comment by paul-brackin** [comment]
Hey @piyush-bansal1989  - I am a new engineer here at Aerospike & I will be helping @khaf on this one.  

Just to update you, I have been able to reproduce your problem.  You were correct in your assessment that there is an incompatibility between the boolean encoding of these 2 clients.  Essentially, the old client is doing it wrong.  I am looking to see if we can get a possible workaround into the new PHP client.  RN it doesn't look like repairing the old client is in the cards.  So, say we got a patch into the new client - you would just have to start using the new client for everything & get rid of the old client.

**Comment by khaf** [comment]
@piyush-bansal1989 We have addressed the issue and will release the fix. It took much longer than anticipated due to the changes required in the Go client, in which we had to come to an agreement regarding the introduction of the support for the old PHP client.

**Comment by piyush-bansal1989** [comment]
@khaf :  Will you please confirm when we get the updated release?

**Comment by paul-brackin** [comment]
@piyush-bansal1989 v1.2.0 has been released

**Comment by piyushbansal0913** [comment]
@paul-brackin : Could you please confirm the steps I need to follow for the new release?

**Comment by paul-brackin** [comment]
@piyush-bansal1989 Yes, you just run one of the 2 new build scripts.  Check out the updated README and let me know if you have specific questions.
